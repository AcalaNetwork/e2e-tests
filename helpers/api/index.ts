import { ApiPromise } from '@polkadot/api'

export const xtokens = {
  relaychainV2: (acc: any) => ({
    V1: {
      parents: 1,
      interior: {
        X1: {
          AccountId32: {
            network: 'Any',
            id: acc,
          },
        },
      },
    },
  }),
  relaychainV3: (acc: any) => ({
    V3: {
      parents: 1,
      interior: {
        X1: {
          AccountId32: {
            id: acc,
          },
        },
      },
    },
  }),
  parachainV2: (paraId: number) => (acc: any) => ({
    V1: {
      parents: 1,
      interior: {
        X2: [
          { Parachain: paraId },
          {
            AccountId32: {
              network: 'Any',
              id: acc,
            },
          },
        ],
      },
    },
  }),
  parachainAccountId20V2: (paraId: number) => (acc: any) => ({
    V1: {
      parents: 1,
      interior: {
        X2: [
          { Parachain: paraId },
          {
            AccountKey20: {
              network: 'Any',
              key: acc,
            },
          },
        ],
      },
    },
  }),
  parachainV3: (paraId: number) => (acc: any) => ({
    V3: {
      parents: 1,
      interior: {
        X2: [
          { Parachain: paraId },
          {
            AccountId32: {
              network: 'Any',
              id: acc,
            },
          },
        ],
      },
    },
  }),
  transfer:
    (token: any, amount: any, dest: (dest: any) => any, weight: any = 'Unlimited') =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      api.tx.xTokens.transfer(token, amount, dest(acc), weight),
  transferMulticurrencies:
    (token: any, amount: any, feeToken: any, feeAmount: any, dest: (dest: any) => any) =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      api.tx.xTokens.transferMulticurrencies(
        [
          [token, amount],
          [feeToken, feeAmount],
        ],
        1,
        dest(acc),
        'Unlimited'
      ),
}

export const xcmPallet = {
  parachainV2: (parents: number, paraId: number) => ({
    V1: {
      parents,
      interior: {
        X1: { Parachain: paraId },
      },
    },
  }),
  relaychainV3: (acc: any) => ({
    V3: {
      parents: 1,
      interior: {
        X1: {
          AccountId32: {
            network: 'Any',
            id: acc,
          },
        },
      },
    },
  }),
  parachainV3: (parents: number, paraId: any) => ({
    V3: {
      parents,
      interior: {
        X1: { Parachain: paraId },
      },
    },
  }),
  limitedReserveTransferAssetsV2:
    (token: any, amount: any, dest: any) =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      (api.tx.xcmPallet || api.tx.polkadotXcm).limitedReserveTransferAssets(
        dest,
        {
          V1: {
            parents: 0,
            interior: {
              X1: {
                AccountId32: {
                  network: 'Any',
                  id: acc,
                },
              },
            },
          },
        },
        {
          V1: [
            {
              id: token,
              fun: { Fungible: amount },
            },
          ],
        },
        0,
        'Unlimited'
      ),
  limitedReserveTransferAssetsV3:
    (token: any, amount: any, dest: any) =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      (api.tx.xcmPallet || api.tx.polkadotXcm).limitedReserveTransferAssets(
        dest,
        {
          V3: {
            parents: 0,
            interior: {
              X1: {
                AccountId32: {
                  id: acc,
                },
              },
            },
          },
        },
        {
          V3: [
            {
              id: token,
              fun: { Fungible: amount },
            },
          ],
        },
        0,
        'Unlimited'
      ),
}

export const tx = {
  xtokens,
  xcmPallet,
}

export const query = {
  balances: ({ api }: { api: ApiPromise }, address: string) => api.query.system.account(address),
  tokens:
    (token: any) =>
    ({ api }: { api: ApiPromise }, address: string) =>
      api.query.tokens.accounts(address, token),
  assets:
    (token: number | bigint) =>
    ({ api }: { api: ApiPromise }, address: string) =>
      api.query.assets.account(token, address),
  evm:
    (contract: string, slot: string) =>
    ({ api }: { api: ApiPromise }, _address: string) =>
      api.query.evm.accountStorages(contract, slot),
}
