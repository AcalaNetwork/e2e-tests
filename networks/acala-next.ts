import { Config } from './types'

import acalaConfig, { Vars } from './acala'

export default {
  ...acalaConfig,
  polkadot: {
    ...acalaConfig.polkadot,
    name: 'acalaNext' as const,
  },
  kusama: {
    ...acalaConfig.kusama,
    name: 'karuraNext' as const,
  },
  config: (opt) => ({
    ...acalaConfig.config(opt),
    options: {
      wasmOverride: {
        polkadot: './wasm/acala_runtime.wasm',
        kusama: './wasm/karura_runtime.wasm',
      }[opt.network],
    },
  }),
} satisfies Config<Vars>
