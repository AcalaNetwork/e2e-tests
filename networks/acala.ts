import { Config } from './types'

export type Vars = {
  relayToken: string
  relayLiquidToken: string
}

export default {
  polkadot: {
    name: 'acala' as const,
    endpoint: 'wss://acala-rpc-0.aca-api.network',
    relayToken: 'DOT',
    relayLiquidToken: 'LDOT',
  },
  kusama: {
    name: 'karura' as const,
    endpoint: 'wss://karura-rpc-0.aca-api.network',
    relayToken: 'KSM',
    relayLiquidToken: 'LKSM',
  },
  config: ({ alice, relayToken, relayLiquidToken }) => ({
    storages: {
      System: {
        account: [[[alice.address], { data: { free: 10 * 1e12 } }]],
      },
      Tokens: {
        accounts: [
          [[alice.address, { Token: relayToken }], { free: 10 * 1e12 }],
          [[alice.address, { Token: relayLiquidToken }], { free: 100 * 1e12 }],
        ],
      },
      Sudo: {
        key: alice.address,
      },
      Homa: {
        // avoid impact test outcome
        $removePrefix: ['redeemRequests', 'unbondings', 'toBondPool'],
        // so that bump era won't trigger unbond
        relayChainCurrentEra: '0x64000000',
      },
      EvmAccounts: {
        accounts: [[['0x82a258cb20e2adb4788153cd5eb5839615ece9a0'], alice.address]],
        evmAddresses: [[[alice.address], '0x82a258cb20e2adb4788153cd5eb5839615ece9a0']],
      },
      PolkadotXcm: {
        $removePrefix: ['versionNotifyTargets', 'versionNotifiers', 'supportedVersion'],
      },
    },
  }),
} satisfies Config<Vars>

export const acala = {
  paraId: 2000,
  paraAccount: '13YMK2eYoAvStnzReuxBjMrAvPXmmdsURwZvc62PrdXimbNy',
  dot: { Token: 'DOT' },
  dai: { Erc20: '0x54a37a01cd75b616d63e0ab665bffdb0143c52ae' },
  wbtc: { ForeignAsset: 5 },
} as const

export const karura = {
  paraId: 2000,
  paraAccount: '13YMK2eYoAvStnzReuxBjMrAvPXmmdsURwZvc62PrdXimbNy',
  ksm: { Token: 'KSM' },
  usdt: { ForeignAsset: 7 },
  dai: { Erc20: '0x4bb6afb5fa2b07a5d1c499e1c3ddb5a15e709a71' },
} as const
