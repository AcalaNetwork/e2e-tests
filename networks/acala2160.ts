import { Config } from './types'

import acalaConfig, { Vars } from './acala'

export default {
  ...acalaConfig,
  polkadot: {
    ...acalaConfig.polkadot,
    name: 'acala2160' as const,
  },
  kusama: {
    ...acalaConfig.kusama,
    name: 'karura2160' as const,
  },
  config: (opt) => ({
    ...acalaConfig.config(opt),
    options: {
      wasmOverride: {
        polkadot: './wasm/acala-2160-dev.wasm',
        kusama: './wasm/karura-2160-dev.wasm',
      }[opt.network],
    },
  }),
} satisfies Config<Vars>
