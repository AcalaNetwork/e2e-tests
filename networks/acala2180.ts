import { Config } from './types'

import acalaConfig, { Vars } from './acala'

export default {
  ...acalaConfig,
  polkadot: {
    ...acalaConfig.polkadot,
    name: 'acala2180' as const,
  },
  kusama: {
    ...acalaConfig.kusama,
    name: 'karura2180' as const,
  },
  config: (opt) => ({
    ...acalaConfig.config(opt),
    options: {
      wasmOverride: {
        polkadot: './wasm/acala-2180.wasm',
        kusama: './wasm/karura-2180.wasm',
      }[opt.network],
    },
  }),
} satisfies Config<Vars>
