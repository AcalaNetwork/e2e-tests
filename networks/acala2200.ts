import { Config } from './types'

import acalaConfig, { Vars } from './acala'

export default {
  ...acalaConfig,
  polkadot: {
    ...acalaConfig.polkadot,
    name: 'acala2200' as const,
  },
  kusama: {
    ...acalaConfig.kusama,
    name: 'karura2200' as const,
  },
  config: (opt) => ({
    ...acalaConfig.config(opt),
    options: {
      wasmOverride: {
        polkadot: './wasm/acala-2200.wasm',
        kusama: './wasm/karura-2200.wasm',
      }[opt.network],
    },
  }),
} satisfies Config<Vars>
