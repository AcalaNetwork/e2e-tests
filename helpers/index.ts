import { ApiPromise } from '@polkadot/api'
import { Codec } from '@polkadot/types/types'
import { expect } from 'vitest'

type CodecOrArray = Codec | Codec[]

const processCodecOrArray = (codec: CodecOrArray, fn: (c: Codec) => any) =>
  Array.isArray(codec) ? codec.map(fn) : fn(codec)

const toHuman = (codec: CodecOrArray) => processCodecOrArray(codec, (c) => c?.toHuman?.() ?? c)
const toHex = (codec: CodecOrArray) => processCodecOrArray(codec, (c) => c?.toHex?.() ?? c)
const toJson = (codec: CodecOrArray) => processCodecOrArray(codec, (c) => c?.toJSON?.() ?? c)

export type EventFilter = string | { method: string; section: string }

export type RedactOptions = {
  number?: boolean | number // precision
  hash?: boolean // 32 byte hex
  hex?: boolean // any hex with 0x prefix
  address?: boolean // base58 address
}

export class Checker {
  readonly #value: any
  readonly #pipeline: Array<(value: any) => any> = []

  #format: 'human' | 'hex' | 'json' = 'json'
  #message: string | undefined
  #redactOptions: RedactOptions | undefined

  constructor(value: any, message?: string) {
    this.#value = value
    this.#message = message
  }

  toHuman() {
    this.#format = 'human'
    return this
  }

  toHex() {
    this.#format = 'hex'
    return this
  }

  toJson() {
    this.#format = 'json'
    return this
  }

  message(message: string) {
    this.#message = message
    return this
  }

  filterEvents(...filters: EventFilter[]) {
    this.toHuman()
    this.#pipeline.push((value) => {
      let data = value.map(({ event: { index: _, ...event } }: any) => event)
      if (filters.length > 0) {
        data = data.filter((evt: any) => {
          return filters.some((filter) => {
            if (typeof filter === 'string') {
              return evt.section === filter
            } else if ('method' in filter) {
              const { section, method } = filter
              return evt.section === section && evt.method === method
            }
          })
        })
      }
      return data
    })
    return this
  }

  redact(options: RedactOptions = { number: 2, hash: true }) {
    this.#redactOptions = {
      ...this.#redactOptions,
      ...options,
    }
    return this
  }

  #redact(value: any) {
    if (!this.#redactOptions) {
      return value
    }

    const redactNumber = this.#redactOptions.number === true || typeof this.#redactOptions.number === 'number'
    const precision = redactNumber
      ? typeof this.#redactOptions.number === 'number'
        ? this.#redactOptions.number
        : 0
      : 0
    const redactHash = this.#redactOptions.hash === true
    const redactHex = this.#redactOptions.hex === true
    const redactAddress = this.#redactOptions.address === true

    const processNumber = (value: number) => {
      if (precision > 0) {
        const rounded = parseFloat(value.toPrecision(precision))
        if (rounded === value) {
          return rounded
        }
        return `(rounded ${rounded})`
      }
      return '(number)'
    }

    const process = (obj: any): any => {
      if (obj == null) {
        return obj
      }
      if (Array.isArray(obj)) {
        return obj.map(process)
      }
      if (redactNumber && typeof obj === 'number') {
        return processNumber(obj)
      }
      if (typeof obj === 'string') {
        if (redactNumber && obj.match(/0x000000[0-9a-f]{26}/)) {
          // this is very likely u128 encoded in hex
          const num = parseInt(obj)
          return processNumber(num)
        }
        if (redactHash && obj.match(/0x[0-9a-f]{64}/)) {
          return '(hash)'
        }
        if (redactHex && obj.match(/0x[0-9a-f]+/)) {
          return '(hex)'
        }
        if (redactAddress && obj.match(/^[1-9A-HJ-NP-Za-km-z]{46,48}$/)) {
          return '(address)'
        }
        if (redactNumber && obj.match(/^-?[\d,]+$/)) {
          const num = parseInt(obj.replace(/,/g, ''))
          return processNumber(num)
        }
        return obj
      }
      if (typeof obj === 'object') {
        return Object.fromEntries(Object.entries(obj).map(([k, v]) => [k, process(v)]))
      }
      return obj
    }

    return process(value)
  }

  map(fn: (value: any) => any) {
    this.#pipeline.push(fn)
    return this
  }

  pipe(fn?: (value: Checker) => Checker) {
    return fn ? fn(this) : this
  }

  async value() {
    let value = await this.#value

    switch (this.#format) {
      case 'human':
        value = toHuman(value)
        break
      case 'hex':
        value = toHex(value)
        break
      case 'json':
        value = toJson(value)
        break
    }

    for (const fn of this.#pipeline) {
      value = await fn(value)
    }

    value = this.#redact(value)

    return value
  }

  async toMatchSnapshot(msg?: string) {
    return expect(await this.value()).toMatchSnapshot(msg ?? this.#message)
  }
}

export const check = (value: any, msg?: string) => {
  if (value instanceof Checker) {
    if (msg) {
      return value.message(msg)
    }
    return value
  }
  return new Checker(value, msg)
}

type Api = { api: ApiPromise }

export const checkEvents = ({ events }: { events: Promise<Codec[] | Codec> }, ...filters: EventFilter[]) =>
  check(events, 'events')
    .filterEvents(...filters)
    .redact()

export const checkSystemEvents = ({ api }: Api, ...filters: EventFilter[]) =>
  check(api.query.system.events(), 'system events')
    .filterEvents(...filters)
    .redact()

export const checkUmp = ({ api }: Api) =>
  check(api.query.parachainSystem.upwardMessages(), 'ump').map((value) =>
    api.createType('Vec<XcmVersionedXcm>', value).toJSON()
  )

export const checkHrmp = ({ api }: Api) =>
  check(api.query.parachainSystem.hrmpOutboundMessages(), 'hrmp').map((value) =>
    (value as any[]).map(({ recipient, data }) => ({
      data: api.createType('(XcmpMessageFormat, XcmVersionedXcm)', data),
      recipient,
    }))
  )
