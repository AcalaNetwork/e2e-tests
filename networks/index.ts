import { SetupOption, setupContext } from '@acala-network/chopsticks-testing'
import { connectParachains, connectVertical } from '@acala-network/chopsticks'
import dotenv from 'dotenv'

import { testingPairs } from '../helpers'

import { Config, Context, NetworkKind } from './types'
import networkDefs from './all'

dotenv.config()

const toNumber = (value: string | undefined): number | undefined => {
  if (value === undefined) {
    return undefined
  }

  return Number(value)
}

export type Network = Awaited<ReturnType<typeof setupContext>> & {
  options: SetupOption
  config: (typeof networkDefs)[keyof typeof networkDefs][NetworkKind]
}
export type NetworkNames = (typeof networkDefs)[keyof typeof networkDefs][NetworkKind]['name']

export const networkCreator = {} as Record<
  NetworkNames,
  (options?: Partial<SetupOption>) => (ctx: Context) => Promise<Network>
>

const relaychains = ['polkadot', 'kusama'] as const

for (const def of Object.values(networkDefs)) {
  for (const relaychain of relaychains) {
    const config = def[relaychain]
    const { endpoint, name } = config
    const upperName = name.toUpperCase()
    networkCreator[name] = (options?: Partial<SetupOption>) => async (ctx: Context) => {
      const setupConfig = (def as Config).config({
        network: relaychain,
        ...config,
        ...ctx,
      })

      const finalOptions: SetupOption = {
        timeout: 600000,
        wasmOverride: process.env[`${upperName}_WASM`],
        blockNumber: toNumber(process.env[`${upperName}_BLOCK_NUMBER`]),
        endpoint: process.env[`${upperName}_ENDPOINT`] ?? endpoint,
        db: process.env.DB_PATH,
        ...setupConfig.options,
        ...options,
      }

      const network = await setupContext(finalOptions)

      await network.dev.setStorage(setupConfig.storages)

      return {
        ...network,
        config,
        options: finalOptions,
      }
    }
  }
}

export const createContext = () => testingPairs()

export const createNetworks = async (
  networkOptions: Partial<Record<NetworkNames, Partial<SetupOption> | undefined>>,
  context = createContext()
) => {
  const ret = {} as Record<NetworkNames, Network>

  let wasmOverriden = false

  for (const [name, options] of Object.entries(networkOptions) as [NetworkNames, Partial<SetupOption> | undefined][]) {
    ret[name] = await networkCreator[name](options)(context)
    wasmOverriden ||= !!ret[name].options.wasmOverride
  }

  const relaychainName = Object.keys(ret).filter(
    (x) => x.startsWith('polkadot') || x.startsWith('kusama')
  )[0] as NetworkKind
  const { [relaychainName]: relaychain, ...parachains } = ret

  if (relaychain) {
    for (const parachain of Object.values(parachains)) {
      await connectVertical(relaychain.chain, parachain.chain)
    }
  }

  const parachainList = Object.values(parachains).map((i) => i.chain)
  if (parachainList.length > 0) {
    await connectParachains(parachainList)
  }

  if (wasmOverriden) {
    // trigger runtime upgrade if needed (due to wasm override)
    for (const chain of Object.values(ret)) {
      await chain.dev.newBlock()
    }
    // handle xcm version message if needed (due to wasm override triggered xcm version upgrade)
    for (const chain of Object.values(ret)) {
      await chain.dev.newBlock()
    }
  }

  return ret
}

// to be compatible with old code

const networks = {} as Record<NetworkNames, (options?: Partial<SetupOption>) => Promise<Network>>

for (const [name, creator] of Object.entries(networkCreator)) {
  networks[name as NetworkNames] = async (options?: Partial<SetupOption>) => creator(options)(createContext())
}

export default networks
