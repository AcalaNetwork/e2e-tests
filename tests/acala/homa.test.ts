import { afterAll, beforeEach, describe, it } from 'vitest'
import { connectVertical } from '@acala-network/chopsticks'
import { sendTransaction, testingPairs } from '@acala-network/chopsticks-testing'

import { checkEvents, checkSystemEvents, checkUmp } from '../../helpers'
import { forceBumpCurrentEra, mint, requestRedeem, sudo } from '../../helpers/api/extrinsics'
import networks, { Network } from '../../networks'

describe('Karura <-> Kusama', async () => {
  let kusama: Network
  let karura: Network

  const { alice } = testingPairs()

  beforeEach(async () => {
    kusama = await networks.kusama({
      wasmOverride: './wasm/kusama_runtime-v9381.compact.compressed.wasm',
    })
    karura = await networks.karura()
    await connectVertical(kusama.chain, karura.chain)

    await karura.dev.setStorage({
      System: {
        Account: [[[alice.address], { data: { free: 10 * 1e12 } }]],
      },
      Tokens: {
        Accounts: [
          [[alice.address, { Token: 'KSM' }], { free: 10 * 1e12 }],
          [[alice.address, { Token: 'LKSM' }], { free: 100 * 1e12 }],
        ],
      },
      Sudo: {
        Key: alice.address,
      },
      Homa: {
        // avoid impact test outcome
        $removePrefix: ['redeemRequests', 'unbondings', 'toBondPool'],
        // so that bump era won't trigger unbond
        relayChainCurrentEra: '0x64000000',
      },
    })
    await kusama.dev.setStorage({
      System: {
        Account: [[[alice.address], { data: { free: 10 * 1e12 } }]],
      },
      ParasDisputes: {
        // those can makes block building super slow
        $removePrefix: ['disputes'],
      },
    })

    // new block to run migration
    await kusama.dev.newBlock()
    await karura.dev.newBlock()

    return async () => {
      await kusama.teardown()
      await karura.teardown()
    }
  })

  it('Homa stake works', async () => {
    const tx0 = await sendTransaction(mint(karura.api, '1000000000000').signAsync(alice, { nonce: 0 }))
    const tx1 = await sendTransaction(
      sudo(karura.api, forceBumpCurrentEra(karura.api, '0')).signAsync(alice, { nonce: 1 })
    )

    await karura.chain.newBlock()

    await checkEvents(tx0, 'homa').toMatchSnapshot()
    await checkEvents(tx1, { section: 'homa', method: 'CurrentEraBumped' }).toMatchSnapshot()
    await checkUmp(karura).redact({ number: true, hex: true }).toMatchSnapshot()

    await kusama.chain.newBlock()

    await checkSystemEvents(kusama, 'ump', 'staking').redact({ address: true, number: true }).toMatchSnapshot()
  })

  it('Homa redeem unbond works', async () => {
    const tx0 = await sendTransaction(requestRedeem(karura.api, '1000000000000', false).signAsync(alice, { nonce: 0 }))
    const tx1 = await sendTransaction(
      sudo(karura.api, forceBumpCurrentEra(karura.api, '0')).signAsync(alice, { nonce: 1 })
    )

    await karura.chain.newBlock()

    await checkEvents(tx0, { section: 'homa', method: 'RequestedRedeem' }).toMatchSnapshot()
    await checkEvents(tx1, { section: 'homa', method: 'RedeemedByUnbond' }).toMatchSnapshot()

    await kusama.chain.newBlock()

    await checkSystemEvents(kusama, 'ump', 'staking').redact({ address: true }).toMatchSnapshot()
  })
})

describe('Karura 3752729 <-> Kusama 16732970', async () => {
  const kusama = await networks.kusama({
    wasmOverride: './wasm/kusama_runtime-v9381.compact.compressed.wasm',
    blockNumber: 16732970,
  })
  const karura = await networks.karura({
    wasmOverride: './wasm/karura-2150.wasm',
    blockNumber: 3752729,
  })
  await connectVertical(kusama.chain, karura.chain)

  const { alice } = testingPairs()

  afterAll(async () => {
    await kusama.teardown()
    await karura.teardown()
  })

  beforeEach(async () => {
    await karura.dev.setStorage({
      System: {
        Account: [[[alice.address], { data: { free: 10 * 1e12 } }]],
      },
      Tokens: {
        Accounts: [
          [[alice.address, { Token: 'KSM' }], { free: 10 * 1e12 }],
          [[alice.address, { Token: 'LKSM' }], { free: 100 * 1e12 }],
        ],
      },
      Sudo: {
        Key: alice.address,
      },
    })
    await kusama.dev.setStorage({
      System: {
        Account: [[[alice.address], { data: { free: 10 * 1e12 } }]],
      },
    })
  })

  it('Homa unbond withdraw works', async () => {
    //Set the relaychain number in advance, Kusama Block Number: 16732970, Karura Block Number: 3752729
    const tx = await sendTransaction(sudo(karura.api, forceBumpCurrentEra(karura.api, '1')).signAsync(alice))
    await karura.chain.newBlock()
    await checkEvents(tx, { section: 'homa', method: 'CurrentEraBumped' }).toMatchSnapshot()

    await kusama.chain.newBlock()
    await checkSystemEvents(kusama, 'ump', 'staking').toMatchSnapshot()
  })
})
