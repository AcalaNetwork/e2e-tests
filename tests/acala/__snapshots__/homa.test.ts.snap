// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Karura <-> Kusama > Homa redeem unbond works > events 1`] = `
[
  {
    "data": {
      "allowFastMatch": false,
      "liquidAmount": 1000000000000,
      "redeemer": "rPizfonc8MPuEsMxufAytHjpRW7a2YrXWHKibaJXe9ZtjKx",
    },
    "method": "RequestedRedeem",
    "section": "homa",
  },
]
`;

exports[`Karura <-> Kusama > Homa redeem unbond works > events 2`] = `
[
  {
    "data": {
      "eraIndexWhenUnbond": "(rounded 4900)",
      "liquidAmount": 1000000000000,
      "redeemer": "rPizfonc8MPuEsMxufAytHjpRW7a2YrXWHKibaJXe9ZtjKx",
      "unbondingStakingAmount": "(rounded 130000000000)",
    },
    "method": "RedeemedByUnbond",
    "section": "homa",
  },
]
`;

exports[`Karura <-> Kusama > Homa redeem unbond works > system events 1`] = `
[
  {
    "data": {
      "amount": "(rounded 130000000000)",
      "stash": "(address)",
    },
    "method": "Unbonded",
    "section": "staking",
  },
  {
    "data": [
      "(hash)",
      {
        "Complete": {
          "proofSize": "(rounded 66000)",
          "refTime": "(rounded 15000000000)",
        },
      },
    ],
    "method": "ExecutedUpward",
    "section": "ump",
  },
]
`;

exports[`Karura <-> Kusama > Homa stake works > events 1`] = `
[
  {
    "data": {
      "liquidAmountAddedToVoid": "(rounded 640000000)",
      "liquidAmountReceived": "(rounded 7600000000000)",
      "minter": "rPizfonc8MPuEsMxufAytHjpRW7a2YrXWHKibaJXe9ZtjKx",
      "stakingCurrencyAmount": 1000000000000,
    },
    "method": "Minted",
    "section": "homa",
  },
]
`;

exports[`Karura <-> Kusama > Homa stake works > events 2`] = `
[
  {
    "data": {
      "newEraIndex": "(rounded 4900)",
    },
    "method": "CurrentEraBumped",
    "section": "homa",
  },
]
`;

exports[`Karura <-> Kusama > Homa stake works > system events 1`] = `
[
  {
    "data": [
      "(hash)",
      {
        "Complete": {
          "proofSize": "(number)",
          "refTime": "(number)",
        },
      },
    ],
    "method": "ExecutedUpward",
    "section": "ump",
  },
  {
    "data": {
      "amount": "(number)",
      "stash": "(address)",
    },
    "method": "Bonded",
    "section": "staking",
  },
  {
    "data": [
      "(hash)",
      {
        "Complete": {
          "proofSize": "(number)",
          "refTime": "(number)",
        },
      },
    ],
    "method": "ExecutedUpward",
    "section": "ump",
  },
]
`;

exports[`Karura <-> Kusama > Homa stake works > ump 1`] = `
[
  {
    "v2": [
      {
        "withdrawAsset": [
          {
            "fun": {
              "fungible": "(number)",
            },
            "id": {
              "concrete": {
                "interior": {
                  "here": null,
                },
                "parents": "(number)",
              },
            },
          },
        ],
      },
      {
        "clearOrigin": null,
      },
      {
        "buyExecution": {
          "fees": {
            "fun": {
              "fungible": "(number)",
            },
            "id": {
              "concrete": {
                "interior": {
                  "here": null,
                },
                "parents": "(number)",
              },
            },
          },
          "weightLimit": {
            "limited": "(number)",
          },
        },
      },
      {
        "depositAsset": {
          "assets": {
            "wild": {
              "all": null,
            },
          },
          "beneficiary": {
            "interior": {
              "x1": {
                "accountId32": {
                  "id": "(hex)",
                  "network": {
                    "any": null,
                  },
                },
              },
            },
            "parents": "(number)",
          },
          "maxAssets": "(number)",
        },
      },
    ],
  },
  {
    "v2": [
      {
        "withdrawAsset": [
          {
            "fun": {
              "fungible": "(number)",
            },
            "id": {
              "concrete": {
                "interior": {
                  "here": null,
                },
                "parents": "(number)",
              },
            },
          },
        ],
      },
      {
        "buyExecution": {
          "fees": {
            "fun": {
              "fungible": "(number)",
            },
            "id": {
              "concrete": {
                "interior": {
                  "here": null,
                },
                "parents": "(number)",
              },
            },
          },
          "weightLimit": {
            "unlimited": null,
          },
        },
      },
      {
        "transact": {
          "call": {
            "encoded": "(hex)",
          },
          "originType": "SovereignAccount",
          "requireWeightAtMost": "(number)",
        },
      },
      {
        "refundSurplus": null,
      },
      {
        "depositAsset": {
          "assets": {
            "wild": {
              "all": null,
            },
          },
          "beneficiary": {
            "interior": {
              "x1": {
                "parachain": "(number)",
              },
            },
            "parents": "(number)",
          },
          "maxAssets": "(number)",
        },
      },
    ],
  },
]
`;

exports[`Karura 3752729 <-> Kusama 16732970 > Homa unbond withdraw works > events 1`] = `
[
  {
    "data": {
      "newEraIndex": "(rounded 100)",
    },
    "method": "CurrentEraBumped",
    "section": "homa",
  },
]
`;

exports[`Karura 3752729 <-> Kusama 16732970 > Homa unbond withdraw works > system events 1`] = `[]`;
